instrument/runtime.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package instrument
+
+import (
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber-go/tally"
+)
+
+// StartReportingRuntimeMetrics starts reporting runtime metrics.
+func StartReportingRuntimeMetrics(
+	scope tally.Scope,
+	reportInterval time.Duration,
+) *RuntimeMetricsReporter {
+	runtimeMetricsReporter := NewRuntimeMetricsReporter(scope, reportInterval)
+	runtimeMetricsReporter.Start()
+	return runtimeMetricsReporter
+}
+
+type runtimeMetrics struct {
+	NumGoRoutines   tally.Gauge
+	GoMaxProcs      tally.Gauge
+	MemoryAllocated tally.Gauge
+	MemoryHeap      tally.Gauge
+	MemoryHeapIdle  tally.Gauge
+	MemoryHeapInuse tally.Gauge
+	MemoryStack     tally.Gauge
+	NumGC           tally.Counter
+	GcPauseMs       tally.Timer
+	lastNumGC       uint32
+}
+
+// RuntimeMetricsReporter A struct containing the state of the RuntimeMetricsReporter.
+type RuntimeMetricsReporter struct {
+	scope          tally.Scope
+	reportInterval time.Duration
+	metrics        runtimeMetrics
+	started        bool
+	quit           chan struct{}
+}
+
+// NewRuntimeMetricsReporter Creates a new RuntimeMetricsReporter.
+func NewRuntimeMetricsReporter(
+	scope tally.Scope,
+	reportInterval time.Duration,
+) *RuntimeMetricsReporter {
+	var memstats runtime.MemStats
+	runtime.ReadMemStats(&memstats)
+
+	memoryScope := scope.SubScope("memory")
```

Comment:
Same comment as above, might be worth considering using a `Tagged` subscope here, something like `scope.Tagged(map[string]string{"category": "memory"})`.